### PR TITLE
Implement wasmi backend

### PR DESCRIPTION
**Pull request by ukint-vs**
Needed for https://github.com/gear-tech/gear/pull/1190
@gear-tech/dev

### CODE REVIEW

File: core-backend/wasmi/src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    fn data_size(&self) -> usize {
+        self.0.current_size().0 * WasmPageNumber::size()
+    }
+
+    unsafe fn get_buffer_host_addr_unsafe(&self) -> HostPointer {
+        self.0.direct_access_mut().as_mut().as_mut_ptr() as HostPointer
+    }
+}
+
+/// can't be tested outside the node runtime
```
Comment (gshep) [inline core-backend/wasmi/src/memory.rs:68]: sorry but didn't get why the node runtime is required there.

File: core-backend/wasmi/src/env.rs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+            Ok(())
+        };
+
+        let Runtime {
+            ext,
+            memory,
+            err: trap,
+        } = self.runtime;
+
+        log::debug!("execution res = {:?}", res);
```
Comment (gshep) [inline core-backend/wasmi/src/env.rs:465]: I would add more details to the log:
```suggestion
        log::debug!("WasmiEnvironment::execute result = {res:?}");
```

File: core-backend/wasmi/src/env.rs [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+        let externval = if self.forbidden_funcs.contains(field_name) {
+            self.map
+                .get(&(
+                    "env".as_bytes().to_owned(),
+                    "forbidden".as_bytes().to_owned(),
+                ))
+                .ok_or_else(|| {
+                    log::debug!(
+                        target: "gwasm",
+                        "Export {}:{} not found",
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/src/env.rs:223]: Just realized that it is a `ImportsResolver` so shouldn't we change all log messages to '... import not found ...' in this implementation? )

**Review by gshep — APPROVED** [review]
Good work! :+1:

File: core-backend/wasmi/src/env.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ukint-vs) [reply core-backend/wasmi/src/env.rs:223]: > Just realized that it is a ImportsResolver so shouldn't we change all log messages to '... import not found ...' in this implementation? )

This is an export, if we call it in terms of the wasm module.

File: core-backend/wasmi/src/env.rs [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+        _memory_type: &MemoryDescriptor,
+    ) -> Result<MemoryRef, wasmi::Error> {
+        let key = (
+            module_name.as_bytes().to_owned(),
+            field_name.as_bytes().to_owned(),
+        );
+        let externval = self.map.get(&key).ok_or_else(|| {
+            log::debug!(
+                target: "gwasm",
+                "Export {}:{} not found",
```
[the remark this answers is not in the document]
Comment (ukint-vs) [reply core-backend/wasmi/src/env.rs:279]: Same as above

File: core-backend/wasmi/src/memory.rs (hunk 1 above)
Comment (ukint-vs) [reply core-backend/wasmi/src/memory.rs:68]: mistake

**Comment by clearloop** [comment]
should we close https://github.com/gear-tech/gear/issues/842 in this PR?

**Comment by ukint-vs** [comment]
> should we close #842 also in this PR?

it will be closed in https://github.com/gear-tech/gear/pull/1190

**Pull request merged**